File: platform-nukkit/src/main/java/taboolib/platform/NukkitPlugin.java
package taboolib.platform;

import cn.nukkit.Server;
import cn.nukkit.plugin.PluginBase;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import taboolib.common.LifeCycle;
import taboolib.common.TabooLibCommon;
import taboolib.common.io.Project1Kt;
import taboolib.common.platform.Platform;
import taboolib.common.platform.PlatformSide;
import taboolib.common.platform.Plugin;
import taboolib.common.platform.function.CommonKt;

import java.io.File;

/**
 * TabooLib
 * taboolib.platform.NukkitPlugin
 *
 * @author sky
 * @since 2021/6/26 8:22 下午
 */
@SuppressWarnings({"Convert2Lambda", "FieldCanBeLocal"})
@PlatformSide(Platform.NUKKIT)
public class NukkitPlugin extends PluginBase {

    @Nullable
    private static final Plugin pluginInstance;
    private static NukkitPlugin instance;

    static {
        TabooLibCommon.lifeCycle(LifeCycle.CONST, Platform.NUKKIT);
        pluginInstance = Project1Kt.findImplementation(Plugin.class);
    }

    public NukkitPlugin() {
        instance = this;
        TabooLibCommon.lifeCycle(LifeCycle.INIT);
    }

    @Override
    public void onLoad() {
        TabooLibCommon.lifeCycle(LifeCycle.LOAD);
        if (pluginInstance != null && !TabooLibCommon.isStopped()) {
            pluginInstance.onLoad();
        }
    }

    @Override
    public void onEnable() {
        TabooLibCommon.lifeCycle(LifeCycle.ENABLE);
        if (pluginInstance != null && !TabooLibCommon.isStopped()) {
            pluginInstance.onEnable();
            Server.getInstance().getScheduler().scheduleTask(this, new Runnable() {
                @Override
                public void run() {
                    TabooLibCommon.lifeCycle(LifeCycle.ACTIVE);
                    pluginInstance.onActive();
                }
            });
        }
        CommonKt.startExecutor();
    }

    @Override
    public void onDisable() {
        TabooLibCommon.lifeCycle(LifeCycle.DISABLE);
        if (pluginInstance != null && !TabooLibCommon.isStopped()) {
            pluginInstance.onDisable();
        }
    }

    @Override
    public File getFile() {
        return super.getFile();
    }

    @NotNull
    public static NukkitPlugin getInstance() {
        return instance;
    }

    @Nullable
    public static Plugin getPluginInstance() {
        return pluginInstance;
    }
}

File: platform-nukkit/src/main/kotlin/taboolib/platform/NukkitPlugin.kt
package taboolib.platform

import cn.nukkit.plugin.PluginBase
import taboolib.common.TabooLibCommon
import taboolib.common.io.findInstance
import taboolib.common.platform.startExecutor
import taboolib.common.platform.submit
import taboolib.plugin.Plugin
import java.io.File

/**
 * TabooLib
 * taboolib.platform.NukkitPlugin
 *
 * @author sky
 * @since 2021/6/14 11:10 下午
 */
class NukkitPlugin : PluginBase() {

    val pluginInstance: Plugin?

    init {
        TabooLibCommon.init()
        instance = this
        pluginInstance = findInstance(Plugin::class.java)
    }

    override fun onLoad() {
        pluginInstance?.onLoad()
    }

    override fun onEnable() {
        pluginInstance?.onEnable()
        startExecutor()
        submit {
            pluginInstance?.onActive()
        }
    }

    override fun onDisable() {
        pluginInstance?.onDisable()
        TabooLibCommon.cancel()
    }

    public override fun getFile(): File {
        return super.getFile()
    }


    companion object {

        lateinit var instance: NukkitPlugin
            private set
    }

}